src/OpenTelemetry/Metrics/MetricAggregators/SumMetricAggregator.cs

```diff
@@ -0,0 +1,248 @@
+
+        private SumMetricDouble sumMetricDouble;
+        private double sumDouble = 0;
+        private double lastSumDouble = 0;
+
+        private DateTimeOffset startTimeExclusive;
+        private DateTimeOffset lastStartTimeExclusive;
+
+        private bool isDeltaValue;
+        private bool isMonotonicValue;
+
+        internal SumMetricAggregator(string name, string description, string unit, Meter meter, DateTimeOffset startTimeExclusive, KeyValuePair<string, object>[] attributes, bool isDeltaValue, bool isMonotonicValue)
+        {
+            this.startTimeExclusive = startTimeExclusive;
+            this.lastStartTimeExclusive = startTimeExclusive;
+            this.sumMetricLong = new SumMetricLong(name, description, unit, meter, startTimeExclusive, attributes);
+            this.sumMetricDouble = new SumMetricDouble(name, description, unit, meter, startTimeExclusive, attributes);
+            this.isDeltaValue = isDeltaValue;
+            this.isMonotonicValue = isMonotonicValue;
+        }
+
+        public void Update<T>(T value)
+            where T : struct
+        {
+            long val = 0;
+            double dval = 0;
+            bool isDoubleValue = false;
+
+            // TODO: Confirm the following casts doesn't cause boxing.
+            // VL: It is expected the Compiler will optimize this heavily!
+            if (typeof(T) == typeof(long))
+            {
+                val = (long)(object)value;
+            }
+            else if (typeof(T) == typeof(int))
+            {
+                val = (int)(object)value;
+            }
+            else if (typeof(T) == typeof(short))
+            {
+                val = (short)(object)value;
+            }
+            else if (typeof(T) == typeof(byte))
+            {
+                val = (byte)(object)value;
+            }
+            else if (typeof(T) == typeof(double))
+            {
+                dval = (double)(object)value;
+                isDoubleValue = true;
+            }
+            else if (typeof(T) == typeof(float))
+            {
+                dval = (float)(object)value;
+                isDoubleValue = true;
+            }
+            else
+            {
+                throw new Exception("Unsupported Type");
+            }
+
+            if (this.isMonotonicValue && (val < 0 || dval < 0))
+            {
+                // Ignore Non-Monotonic Values
+                // TODO: Log?
+            }
+            else
+            {
+                if (this.isDeltaValue)
+                {
+                    // Updating with a Delta Value
+
+                    if (isDoubleValue)
+                    {
+                        if (dval != 0)
+                        {
+                            lock (this.lockUpdate)
+                            {
+                                // TODO: Replace Lock with Interlocked.Add
+                                // VL: We have multiple fields to lock/protect,
```

Comment:
will discuss separately - i only see one value to be updated - either a double or a long. so interlocked for long, and try-spin-add for double should work. (will address this separate PR)